File: _dsv/fromhtml.py
import sys
from html.parser import HTMLParser
import html
import argparse
import threading
from queue import Queue
from concurrent.futures import Future
from functools import partial
from ._base import _Base

class Parser(HTMLParser):
    def __init__(self, callback, strict=False, inner_html=False):
        super().__init__()
        self.callback = callback
        self.state = []
        self.got_header = False
        self.strict = strict
        self.inner_html = inner_html
        self.rowspans = {}

    def apply_rowspans(self):
        i = len(self.current_row) + 1
        while value := self.rowspans.get(i):
            self.current_row.append(value[1])
            i += 1

    def decrement_rowspans(self):
        for k, v in list(self.rowspans.items()):
            if v[0] <= 1:
                self.rowspans.pop(k)
            else:
                v[0] -= 1

    def handle_starttag(self, tag, attrs):
        if self.state and self.state[-1] in {'th', 'td'}:
            if self.inner_html:
                self.current_row[-1] += f'<{tag} {' '.join(f'{k}="{html.escape(v)}"' for k, v in attrs)}>'
            return

        self.state.append(tag)
        match self.state[-2:]:
            case \
                  ['table' | 'thead' | 'tbody'] \
                | ['table', 'thead' | 'tbody' | 'tr'] \
                | ['thead' | 'tbody', 'tr'] \
                | ['tr', 'th' | 'td']:
                # good

                if tag == 'tr':
                    # new row
                    self.current_row = []
                    self.decrement_rowspans()
                    self.apply_rowspans()

                elif tag in {'td', 'th'}:
                    self.apply_rowspans()
                    # new column
                    self.current_row.append('')

                    if rowspan := dict(attrs).get('rowspan'):
                        if rowspan.isnumeric() and (rowspan := int(rowspan)) and rowspan > 0:
                            self.rowspans[len(self.current_row)] = [rowspan, '']
                        else:
                            print(f'invalid rowspan {rowspan!r}', file=sys.stderr)

                else:
                    # new table
                    self.rowspans.clear()

            case _:
                # bad
                if self.strict:
                    raise ValueError(f'invalid tags {self.state}')
                else:
                    self.state.pop()

    def handle_endtag(self, tag):
        old_state = self.state

        if self.state and self.state[-1] in {'th', 'td'} and tag != self.state[-1]:
            if self.inner_html:
                self.current_row[-1] += f'</{tag}>'

        if tag in self.state:
            ix = list(reversed(self.state)).index(tag) + 1
            self.state = self.state[:-ix]

        if 'tr' in old_state and 'tr' not in self.state:
            is_header = 'thead' in old_state
            if is_header and self.got_header:
                print('got duplicate html table header', file=sys.stderr)
            else:
                self.apply_rowspans()
                self.callback((self.current_row, is_header))
                self.got_header = is_header

    def handle_data(self, data):
        if self.state and self.state[-1] in {'td', 'th'}:
            self.current_row[-1] += data

            if rowspan := self.rowspans.get(len(self.current_row)):
                rowspan[1] = self.current_row[-1]

class fromhtml(_Base):
    ''' convert from html table '''
    parser = argparse.ArgumentParser()
    parser.add_argument('--strict', action='store_true', help='only allow valid table')
    parser.add_argument('--inner-html', action='store_true', help='output the innerHTML of table cells, not the innerText')

    def process_file(self, file, do_callbacks=True, do_yield=False):
        self.determine_delimiters(b'')

        got_row = False
        queue = Queue()
        thread = threading.Thread(target=self.parse, args=[queue, file])
        thread.start()

        while True:
            fut = queue.get()
            if do_callbacks and (fut is None or fut.exception() is not None):
                self.on_eof()
            if fut is None:
                break

            got_row = True
            item = fut.result()
            row, is_header = item
            row = [x.encode('utf8').strip() for x in row]
            if do_callbacks and (self.on_header(row) if is_header else self.on_row(row)):
                break
            if do_yield:
                yield item

        thread.join()
        return got_row

    def parse(self, queue, *args, **kwargs):
        try:
            self._parse(queue, *args, **kwargs)
        except Exception as e:
            fut = Future()
            fut.set_exception(e)
            queue.put_nowait(fut)
        finally:
            queue.put_nowait(None)

    def parser_callback(self, queue, result):
        fut = Future()
        fut.set_result(result)
        queue.put_nowait(fut)

    def _parse(self, queue, file, chunk=8192):
        parser = Parser(partial(self.parser_callback, queue), self.opts.strict, self.opts.inner_html)
        remainder = b''
        while buf := file.read1(chunk):
            buf = remainder + buf

            try:
                utf8_buf = buf.decode('utf8')
            except UnicodeDecodeError as e:
                utf8_buf = buf[:e.start].decode('utf8')
                if not utf8_buf:
                    raise
                remainder = buf[e.start:]
            else:
                remainder = b''

            parser.feed(utf8_buf)

        if remainder:
            # probably raises a decode error
            remainder.decode('utf8')

        parser.close()
        queue.put_nowait(None)

File: src/subcommands/fromhtml.rs
use anyhow::Result;
use crate::base;
use std::io::BufRead;
use std::process::ExitCode;
use std::collections::HashMap;
use bstr::{BString, ByteVec};
use clap::{Parser, ArgAction};
use quick_xml::{events::Event, reader::Reader, errors::SyntaxError};

#[derive(Parser, Default)]
#[command(about = "convert from html table")]
pub struct Opts {
    #[arg(long, action = ArgAction::SetTrue, help = "only allow valid table")]
    strict: bool,
    #[arg(long, action = ArgAction::SetTrue, help = "output the innerHTML of table cells, not the innerText")]
    inner_html: bool,
}

pub struct Handler {
    opts: Opts,
}

impl Handler {
    pub fn new(opts: Opts, _base: &mut base::Base, _is_tty: bool) -> Result<Self> {
        Ok(Self {
            opts,
        })
    }
}

type Rowspans = HashMap<usize, (usize, BString)>;

fn apply_rowspans(row: &mut Vec<BString>, rowspans: &Rowspans) {
    let mut i = row.len() + 1;
    while let Some((_, value)) = rowspans.get(&i) {
        row.push(value.clone());
        i += 1;
    }
}

fn decrement_rowspans(rowspans: &mut Rowspans) {
    let _ = rowspans.extract_if(|_, (span, _)| {
        *span = span.saturating_sub(1);
        *span == 0
    });
}

fn add_rowspan(rowspans: &mut Rowspans, column: usize, span: usize, value: BString) {
    rowspans.insert(column, (span, value));
}

impl base::Processor for Handler {

    fn process_file<R: BufRead>(mut self, file: R, base: &mut base::Base, do_callbacks: base::Callbacks) -> Result<ExitCode> {

        let ofs = self.determine_delimiters(b"".into(), &base.opts).1;
        if base.on_ofs(ofs) {
            return Ok(ExitCode::SUCCESS)
        }

        let mut state: Vec<BString> = vec![];
        let mut current_row: Vec<BString> = vec![];
        let mut got_header = false;
        let mut buffer = vec![];
        let mut rowspans = Rowspans::new();

        let mut reader = Reader::from_reader(file);
        let config = reader.config_mut();
        config.allow_dangling_amp = true;
        config.allow_unmatched_ends = true;
        config.check_end_names = false;
        config.expand_empty_elements = true;

        loop {
            buffer.clear();
            match reader.read_event_into(&mut buffer) {
                Ok(Event::Start(tag)) => {
                    if matches!(state.last().map(|x| x.as_slice()), Some(b"td" | b"th")) {
                        if self.opts.inner_html && let Some(last) = current_row.last_mut() {
                            last.push(b'<');
                            last.push_str(&*tag);
                            last.push(b'>');
                        }
                        continue
                    }

                    let name = tag.local_name();
                    let name = name.as_ref();
                    match (state.last().map(|x| x.as_slice()), name) {
                        (None, b"table" | b"thead" | b"tbody")
                        | (Some(b"table"), b"thead" | b"tbody" | b"tr")
                        | (Some(b"thead" | b"tbody"), b"tr")
                        | (Some(b"tr"), b"th" | b"td")
                        => {
                            state.push(name.into());
                            match name {
                                // good
                                b"tr" => {
                                    // new row
                                    current_row.clear();
                                    decrement_rowspans(&mut rowspans);
                                    apply_rowspans(&mut current_row, &rowspans);
                                },
                                b"td" | b"th" => {
                                    apply_rowspans(&mut current_row, &rowspans);
                                    // new column
                                    current_row.push(b"".into());
                                    for attr in tag.html_attributes().with_checks(false) {
                                        let attr = attr?;
                                        if attr.key.0 == b"rowspan" {
                                            if let Ok(Ok(span)) = std::str::from_utf8(&attr.value).map(str::parse::<usize>) && span > 0 {
                                                add_rowspan(&mut rowspans, current_row.len(), span, b"".into());
                                            } else {
                                                eprintln!("invalid rowspan {:?}", attr.value);
                                            }
                                        }
                                    }
                                },
                                _ => {
                                    // new table
                                    rowspans.clear();
                                }
                            }
                        },
                        _ => {
                            // bad
                            if self.opts.strict {
                                anyhow::bail!("invalid tags {:?}", state)
                            }
                        },
                    }

                },
                Ok(Event::End(tag)) => {
                    let name = tag.local_name();
                    let name = name.as_ref();
                    match state.last().map(|x| x.as_slice()) {
                        Some(x @ (b"th" | b"td")) if x != name && self.opts.inner_html => {
                            if let Some(last) = current_row.last_mut() {
                                last.push_str(b"</");
                                last.push_str(&*tag);
                                last.push(b'>');
                            }
                        },
                        _ => (),
                    }
                    let had_tr = state.iter().any(|x| x == b"tr");
                    let had_thead = state.iter().any(|x| x == b"thead");

                    if let Some(pos) = state.iter().rposition(|t| t == name) {
                        state.drain(pos..);
                    }

                    if had_tr && !state.iter().any(|x| x == b"tr") {
                        if had_thead && got_header {
                            eprintln!("got duplicate html table header");
                        } else if had_thead && do_callbacks.contains(base::Callbacks::ON_HEADER) {
                            self.on_header(base, current_row.clone())?;
                        } else if !had_thead && do_callbacks.contains(base::Callbacks::ON_HEADER) {
                            self.on_row(base, current_row.clone())?;
                        }
                        got_header = had_thead;
                    }
                },
                Ok(Event::Text(text)) => {
                    let row_len = current_row.len();
                    if matches!(state.last().map(|x| x.as_slice()), Some(b"td" | b"th")) && let Some(last) = current_row.last_mut() {
                        let text = text.into_inner();
                        let len = text.len();
                        last.push_str(text);
                        if let Some((_, value)) = rowspans.get_mut(&row_len) {
                            value.push_str(&last[last.len() - len ..]);
                        }
                    }
                },
                Ok(Event::Eof) => break,
                Err(quick_xml::errors::Error::Syntax(SyntaxError::UnclosedComment | SyntaxError::UnclosedPIOrXmlDecl | SyntaxError::UnclosedDoctype | SyntaxError::UnclosedCData | SyntaxError::UnclosedTag)) => (),

                Ok(_) => (),
                Err(e) => Err(e)?,
            }
        }
        if do_callbacks.contains(base::Callbacks::ON_EOF) {
            self.on_eof(base)?;
        }
        Ok(ExitCode::SUCCESS)
    }
}
